pkg/provision/cleanup.go

```diff
@@ -0,0 +1,250 @@
+	exe := exec.Command(cmd, args...)
+	output, err := exe.CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "execute '%v %+v': %v", cmd, args, string(output))
+	}
+
+	return output, nil
+}
+
+func lines(input []byte) []string {
+	input = bytes.TrimSpace(input)
+	if len(input) == 0 {
+		return nil
+	}
+
+	return strings.Split(string(input), "\n")
+}
+
+func must(input []byte, err error) []byte {
+	if err != nil {
+		panic(err)
+	}
+
+	return input
+}
+
+type container struct {
+	ID   string `json:"ID"`
+	Spec struct {
+		Root struct {
+			Path string `json:"path"`
+		} `json:"root"`
+		Mounts []struct {
+			Source string `json:"source"`
+		} `json:"mounts"`
+	} `json:"Spec"`
+}
+
+// we declare this method as a variable so we can
+// mock it in testing.
+func initZdbConnection(id string) zdb.Client {
+	socket := fmt.Sprintf("unix://%s/zdb.sock", socketDir(id))
+	return zdb.New(socket)
+}
+
+func socketDir(containerID string) string {
+	return fmt.Sprintf("/var/run/zdb_%s", containerID)
+}
+
+// CleanupResources cleans up unused resources
+func CleanupResources(msgBrokerCon string) error {
+	toSave, toDelete, err := checkContainers(msgBrokerCon)
+	if err != nil {
+		return errors.Wrap(err, "failed to check containers")
+	}
+
+	ctx := context.TODO()
+	utils := filesystem.NewUtils()
+	pools, err := ioutil.ReadDir("/mnt")
+	if err != nil {
+		return errors.Wrap(err, "failed to list available pools")
+	}
+
+	for _, pool := range pools {
+		if !pool.IsDir() {
+			continue
+		}
+		path := filepath.Join("/mnt", pool.Name())
+		subvols, err := utils.SubvolumeList(ctx, path)
+		if err != nil {
+			// A Pool might not be mounted here, skip it
+			log.Error().Err(err).Msgf("failed to list available subvolumes in: %s", path)
+			continue
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// if the subvolume is a zdb and has 0 maxRfer, don't skip here. It might need to be deleted
+			zdbVolume := strings.HasPrefix(subvol.Path, "zdb") && qgroup.MaxRfer != 0
+
+			// we only handle volumes that are 256MiB or 10MiB
+			if qgroup.MaxRfer != 268435456 && qgroup.MaxRfer != 10485760 && !zdbVolume {
```

Comment:
Why do we only check volumes that have these specific size ?

Reply:
I can't remember fully remember but the 250MB size is the default "container" rootfs volume so we sure it's created for that purpose. the 10MB check is similar because we use this size for zdb container rootfs.

We were trying to be as defensive as possible with the clean up

Reply:
May be we don't need this anymore. If we make sure the volume is not used by a container or an active volume reservation

Reply:
I would keep it anyway, as you said. Try to be defensive

Reply:
This code was extracted from a script that has a very different purpose. We are now trying to get something the cleans up everything that should not be there.